packages/wrangler/src/pages/functions/routes-validation.ts

```diff
@@ -0,0 +1,172 @@
+import { FatalError } from "../../errors";
+import {
+	MAX_FUNCTIONS_ROUTES_RULES,
+	MAX_FUNCTIONS_ROUTES_RULE_LENGTH,
+	ROUTES_SPEC_VERSION,
+} from "../constants";
+import { getRoutesValidationErrorMessage } from "../errors";
+import type { RoutesJSONSpec } from "./routes-transformation";
+
+export enum RoutesValidationError {
+	INVALID_JSON_SPEC,
+	NO_INCLUDE_RULES,
+	INVALID_RULES,
+	TOO_MANY_RULES,
+	RULE_TOO_LONG,
+	OVERLAPPING_RULES,
+}
+
+/**
+ *  Check if given routes data is a valid RoutesJSONSpec
+ */
+export function isRoutesJSONSpec(data: unknown): data is RoutesJSONSpec {
+	return (
+		(typeof data === "object" &&
+			data &&
+			"version" in data &&
+			typeof (data as RoutesJSONSpec).version === "number" &&
+			(data as RoutesJSONSpec).version === ROUTES_SPEC_VERSION &&
+			Array.isArray((data as RoutesJSONSpec).include) &&
+			Array.isArray((data as RoutesJSONSpec).exclude)) ||
+		false
+	);
+}
+
+export function validateRoutes(routesJSON: RoutesJSONSpec, routesPath: string) {
+	try {
+		if (!isRoutesJSONSpec(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.INVALID_JSON_SPEC,
+					routesPath
+				),
+				1
+			);
+		}
+
+		if (!hasIncludeRules(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.NO_INCLUDE_RULES,
+					routesPath
+				),
+				1
+			);
+		}
+
+		if (!hasValidRulesCount(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.TOO_MANY_RULES,
+					routesPath
+				),
+				1
+			);
+		}
+
+		if (!hasValidRuleCharCount(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.RULE_TOO_LONG,
+					routesPath
+				),
+				1
+			);
+		}
+
+		if (!hasValidRules(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.INVALID_RULES,
+					routesPath
+				),
+				1
+			);
+		}
+
+		if (hasNoOverlappingRules(routesJSON)) {
+			throw new FatalError(
+				getRoutesValidationErrorMessage(
+					RoutesValidationError.OVERLAPPING_RULES,
+					routesPath
+				),
+				1
+			);
+		}
+	} catch (err) {
+		// TODO what do we want here
```

Comment:
I'd think consumers of `validateRoutes` would handle these errors if necessary. No need to try/catch

Reply:
👍